### ISSUE TITLE
Add support for DRµGS

### ISSUE THREAD

**Issue by Chanka0** [ggerganov/llama.cpp#4704]
# Feature Description

[DRµGS](https://github.com/EGjoni/DRUGS) directly injects noise into transformer layers at inference time to alter the outcomes of predictions as an alternative to conventional sampling methods. The author has already given an in-depth description in their repo.

# Motivation

It's a unique and highly customizable approach to modifying the outcome from sampling. From my understanding it would act as an alternative to normal sampling methods, as adding samplers atop of this implementation would reduce its effectiveness.

# Possible Implementation

[The repo](https://github.com/EGjoni/DRUGS) already details how it is implemented along with source code to accompany it.

**Comment by Alumniminium** [comment]
this might actually be the best approach

**Comment by cryolite-ai** [comment]
This may be a silly point, @Chanka0  but not knowing the background, when I saw the title, I sighed and assumed it was a request for support for some sort of obscure opinionated uncensored LLM model.. competing for attention amongst 800+ other issues in an increasingly overwhelmed project repo. OK, my bad..

..but perhaps it's worth expanding the title to "Transformer layer inference noise injection and sampling (support for DRµGS)" and it might get better profile with those who might want to try it as a sampling methodology (before that strategy gets any wider awareness). In any case, I think you probably need to make a better case why the development should happen within llama.cpp itself as opposed to a project built on it (which, if it demonstrates good utility, might become a PR option if stable and written to conform to llama.cpp structure). Just a thought..

**Comment by EGjoni** [comment]
>  In any case, I think you probably need to make a better case why the development should happen within llama.cpp itself as opposed to a project built on it

The method requires intimate contact with the kv-cache. I'm not super familiar with what sorts of functionality llama.cpp exposes to things built on top of it, but after having spent like 3 days clawing at the guts of the huggingface transformers library, I would not only be impressed to learn that this would be feasible from outside of llama.cpp, but I would even have a little bit of my faith in the world restored.

Anyway, a porting guide is up, if anyone is interested in implementing it (depending of course, on whether or not it should be implemented here):

https://github.com/EGjoni/DRUGS/blob/main/porting/A%20Guide%20to%20Making%20DRUGS.md

**Comment by keturn** [comment]
stalebot, that is not sufficient reason to close an issue as "completed."

I don't think we've seen any evidence of this feature being implemented in this project.

**Comment by Chanka0** [comment]
> stalebot, that is not sufficient reason to close an issue as "completed."
> 
> I don't think we've seen any evidence of this feature being implemented in this project.

I don't believe so either, but I haven't kept too close of a track of the commits. I'm tempted to ping ggerganov but I also don't want to cause any issues so... ¯\\\_(ツ)\_/¯

**Comment by ggerganov** [comment]
The bot closes stale issue that didn't have any activity for a long time. It does not mean it was completed but it also does not mean it is not planned. It just means that no one attempted to move it forward recently.

If there is still interest, feel free to reopen. I believe the author of the issue should be able to do that and also remove the `stale` label?
